### PR TITLE
extract aggregate to input list (SR-8573)

### PR DESCRIPTION
**Pull request by marcrasi**
I wrote most of this up while waiting for the AD ticket, so I thought I might as well finish it and send it out for review!

### CODE REVIEW

**Review by rxwei — COMMENTED** [review]
Nice work. This approach makes high-level sense, but it has a big issue: this cannot produce a list of **one** element when the operand type is `Tensor`. So single value datasets cannot be expressed.

Concretely, when `Dataset` is specialize as `Dataset<Tensor<Float>>`, the input to `#tfop("TensorSliceDataset")` should be a list instead of a single value.

**Comment by rxwei** [comment]
So we need a way to tell the compiler that whether the input should be a single value or a list. Unlike attributes, inputs do not have labels so adding a $-suffix probably won't work.

Instead of unpacking aggregates directly, I would suggest making unpacking work on array literal elements like a flat map operation. It unpacks all elements of an array literal and concatenate the results together.

Here are two examples:
```swift
let x: Tenors<Float>
#tfop("TensorSliceDataset", [x], ...)
// after unpacking:
// #tfop("TensorSliceDataset", [x], ...)
```

```swift
let x: (Tenors<Float>, (Tensor<Float>, Tensor<Float>))
#tfop("TensorSliceDataset", [x], ...)
// after unpacking:
// #tfop("TensorSliceDataset", [x.0, x.1.0, x.1.1], ...)
```

File: test/TensorFlow/tfop_packing.swift [hunk 1]
```diff
@@ -56,6 +56,60 @@ public func packResultsToAggregate_externalStructFixedLayout() {
...
+  return #tfop("SomeOp6", (t, Tensor<Float>(1)))
+}
+
+public func unpackAggregate_generic() {
+  struct Foo {
+    let x: Tensor<Float>
+    let y: (Tensor<Float>, (a: Tensor<Int32>, b: Tensor<Bool>))
+  }
+  let foo = Foo(x: Tensor(1), y: (Tensor(2), (a: Tensor(3), b: Tensor(false))))
+  _hostOp(genericUnpackedAggregate(t: foo))
```
Comment (rxwei) [inline test/TensorFlow/tfop_packing.swift]: It seems unnecessary to call `_hostOp`.
Just `_ = genericUnpackedAggregate(t: foo)` would do.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 2]
```diff
@@ -2148,14 +2149,51 @@ void TFDeabstraction::formGraphOp(SILTensorOpInfo &opInfo,
         continue;
       }
 
-      // Tensor operand type's are ok.
+      // Handle TensorFlow values. These must be handled separately from
+      // aggregates of TensorFlow values (handled below), so that we do not
+      // generate input lists for them.
       if (isTensorFlowValue(operandTy)) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Please also change the non-list cases to auto-unwrap, when the aggregate has exactly one terminal element.

After this is done, all the hacky AST-level unpacking code can be nuked! https://github.com/apple/swift/blob/ed4eab5bd272192c5a44f2de7aa24db49472368d/lib/Sema/CSApply.cpp#L2477 (lines 2477-2527).

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 3]
```diff
@@ -2148,14 +2149,51 @@ void TFDeabstraction::formGraphOp(SILTensorOpInfo &opInfo,
...
+      // Remember that we have an input list, this marker is important because
+      // the list may be empty, and we need to know it exists in graph lowering.
+      opName +=
+        GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_InputList);
+      const char *elementMarker =
+        GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_InputListElt);
+
+      // Handle aggregates of TensorFlow values by unpacking them to an input
+      // list.
+      if (isTensorFlowValueOrAggregate(operandTy.getASTType())) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: `isTensorFlowValueOrAggregate` goes recursively into the aggregate, so it's inefficient to visit the aggregate the second time. How about just recursively try to unpack any aggregate and throwing a diagnostic if there's any non-TF type?

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 4]
```diff
@@ -2148,61 +2166,74 @@ void TFDeabstraction::formGraphOp(SILTensorOpInfo &opInfo,
...
+        // reject this situation.
+        auto astTypeString = operandTy.getASTType().getString();
+        diagnoseInvalid("operand of type '" + astTypeString + "' must be " +
+                        "wrapped in an array because it is an " +
+                        (unpackedInputsCount == 0
+                            ? "empty aggregate"
+                            : "aggregate of more than one TensorFlow value"));
       }
+      opName +=
+        GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_Normal);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Indent by 4

File: lib/Sema/CSApply.cpp [hunk 5]
```diff
@@ -2471,64 +2471,9 @@ namespace {
     /// TensorProtocol (like Tensor or TensorElementLiteral), we use the
     /// TensorHandle that they contain instead.
     Expr *visitTFOp(ObjectLiteralExpr *expr) {
-      auto &ctx = cs.getASTContext();
```
Comment (rxwei) [inline lib/Sema/CSApply.cpp]: This is fantastic.

**Comment by marcrasi** [comment]
I have modified it to do aggregate unpacking inside arrays, and I replaced a bunch of existing unpacking code with the new general unpacking code. This PR removes more non-test lines than it adds!

To get everything to work, I had to add a bunch of fancy features to the general unpacking code, because those fancy features existed in the unpacking code that I removed: `unwrapAggregateInstructions` and `loweredTupleElts`/`loweredStructFields`.

I did not make the new code use `TensorProtocol` like the old code. It doesn't seem necessary right now. If we ever need it, it should be easy to add back in as a recursive base case in `unpackTensorAggregates`.

**Review by rxwei — APPROVED** [review]
This looks great. Removing logic that only extracts the `handle` out of a `TensorProtocol`-conforming type is totally the right move, since unpacking generalizes everything.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**